### ISSUE TITLE
[Question] Quasi-periodic kernel hyperparameters and constant prediction value

### ISSUE THREAD

**Issue by GStechschulte** [cornellius-gp/gpytorch#1909]
Hello, I am attempting to model the power consumption of a machine in which the data is collected in 5 minute intervals. The data exhibits periodicity with local variations and so I have decided in constructing a quasi-periodic kernel to model the underlying function. For example, a snippet of the data (x-axis is 5-minute intervals and y-axis is kilowatts):

![not_zoom](https://user-images.githubusercontent.com/63432018/153145581-b7b2d2a7-0715-4f68-afa6-565259453cc5.png)

To model the periodicity I use the `gpytorch.kernels.PeriodicKernel` and for the local / short term variations I use the `gpytorch.kernels.RBFKernel`. These two kernels are then multiplied. For the first model specification, I don't include any prior distributions on the hyperparameters as I would like to see what the optimization of the marginal log likelihood returns after training.

Model specification with no priors on hyperparameters:

```python
class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)

        self.mean_module = gpytorch.means.ConstantMean()
       
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.PeriodicKernel() * gpytorch.kernels.RBFKernel()
            )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

# initialize likelihood and model
likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = ExactGPModel(X_train, y_train, likelihood)
```

The training loop:

```python
with gpytorch.settings.max_cg_iterations(5000):
    # Use the adam optimizer
    optimizer = torch.optim.Adam([
        {'params': model.parameters()},  # Includes Gaussian Likelihood parameters
    ], lr=0.01)

    # "Loss" for GPs - the marginal log likelihood
    mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

    training_iter = 300
    for i in range(training_iter):
        # Zero gradients from previous iteration
        optimizer.zero_grad()
        # Output from model
        output = model(X_train)
        # Calc loss and backprop gradients
        loss = -mll(output, y_train)
        loss.backward()
        if (i % 20) == 0:
            print('Iter %d/%d - Loss: %.3f noise: %.6f' % (
                i + 1, training_iter, loss.item(),
                model.likelihood.noise.item()
            ))
            print(list(model.parameters()))
        optimizer.step()
```
```
Iter 1/300 - Loss: 2.372 noise: 0.693247
[Parameter containing:
tensor([0.], requires_grad=True), Parameter containing:
tensor([0.], requires_grad=True), Parameter containing:
tensor(0., requires_grad=True), Parameter containing:
tensor([[0.]], requires_grad=True), Parameter containing:
tensor([[0.]], requires_grad=True), Parameter containing:
tensor([[0.]], requires_grad=True)]
Iter 21/300 - Loss: 2.055 noise: 0.793179
[Parameter containing:
tensor([0.1908], requires_grad=True), Parameter containing:
tensor([0.1969], requires_grad=True), Parameter containing:
tensor(0.1908, requires_grad=True), Parameter containing:
tensor([[3.2367e-05]], requires_grad=True), Parameter containing:
tensor([[0.0019]], requires_grad=True), Parameter containing:
tensor([[0.0015]], requires_grad=True)]
Iter 41/300 - Loss: 1.853 noise: 0.873706
. . .
Iter 281/300 - Loss: 0.321 noise: 0.145783
[Parameter containing:
tensor([-1.8526], requires_grad=True), Parameter containing:
tensor([1.8786], requires_grad=True), Parameter containing:
tensor(-1.8526, requires_grad=True), Parameter containing:
tensor([[0.0001]], requires_grad=True), Parameter containing:
tensor([[0.0097]], requires_grad=True), Parameter containing:
tensor([[0.0083]], requires_grad=True)]
```

Posterior Predictive Mode:

```python
# Get into evaluation (predictive posterior) mode
model.eval()
likelihood.eval()

# Make predictions by feeding model through likelihood
with torch.no_grad(), gpytorch.settings.fast_pred_var(), gpytorch.settings.max_cg_iterations(9000):
    #test_x = test_x.to(torch.float32)
    test_pred = likelihood(model(X_test))
    train_pred = likelihood(model(X_train))
```

Plotting training and prediction on test set:

```python
with torch.no_grad():

    # Initialize plot
    f, ax = plt.subplots(1, 1, figsize=(20, 6))
    # Plot training data as black stars
    ax.plot(X_train.numpy(), y_train.numpy(), alpha=0.7)
    ax.plot(X_train.numpy(), train_pred.mean.numpy())
    # Plot predictive means as blue line
    ax.plot(X_test.numpy(), y_test.numpy())
    ax.plot(X_test.numpy(), test_pred.mean.numpy(), lw=4)
    ax.legend(['Observed Data', 'Training', 'Test', 'Predictions'])
```

![series](https://user-images.githubusercontent.com/63432018/153151235-46288e5d-2d71-4f86-8348-9cc2fedc2cfe.png)

The training seems to capture some of the periodicity and short term fluctuations in energy consumption, however, the variation of the periodicity is lacking. This makes me think including an output scale with a prior distribution will help in capturing more of the variability. 

Indeed, the output scale is rather small:

```python
print('Actual outputscale:', model.covar_module.outputscale)
```
```
Actual outputscale: tensor(0.1139, grad_fn=<SoftplusBackward0>)
```

Also, the predictions on the test set are nearly constant. In fact, the predictions on the test set are just the constant mean that was learned in training:

```python
print(output.mean), print(test_pred)
```

results in all elements of both tensors = 1.8919

Why is it that the predictions on the test data are at constant mean of the training data? Nonetheless, to capture more of the variation I then included a output scale prior based on the learned output scale from the previous training and initialized it in the model specification:

```python
class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)

        self.mean_module = gpytorch.means.ConstantMean()
        outputscale_prior = gpytorch.priors.GammaPrior(0.5, 0.15)

        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.PeriodicKernel() * gpytorch.kernels.RBFKernel(),
                outputscale_prior=outputscale_prior
            )

        # Initialize output scale to mean of priors
        self.covar_module.outputscale = outputscale_prior.mean

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

# initialize likelihood and model
likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = ExactGPModel(X_train, y_train, likelihood)
```

Rerunning the training loop and posterior prediction results in the following plot:

![new_preds](https://user-images.githubusercontent.com/63432018/153159052-d7744bd1-08f0-4d44-8c12-ef2a6ed06df3.png)

The prior on the output scale results in more of the variation being captured (which was expected), however the constant prediction value stays the same. Again, the mean of the `test_preds` and `output` are the same. 

My end question is:

1.) Why is my prediction value resulting in just the learned mean during training?
2.) Including the output scale prior in the `ScaleKernel` decorator, this then results in the prior being applied to both kernels. However, if one wants to have a different output scale prior on each kernel, would you scale each kernel individually with each kernel having its own output scale prior and then sum or multiply the kernels?

Other comments:

I have tried changing `ConstantMean()` to `ZeroMean()` and this still did not solve the problem.
I ran the training loop twice, each with 300 iterations, and this also did not solve the problem.
I decreased the learning rate to 0.001 and increased it to 0.1, and this also did not solve the problem. 
I have consulted the hyperparameters notebook in the repo, and still could not find a solution. 

Thank you for your time and help.

**Comment by wjmaddox** [comment]
A couple of possibilities here, but it's a bit tricky to tell without the data although i can probably mock it if necessary: 

- are you running the GP in single or double precision? try running in double.
- try also standardizing the inputs to have range [0, 1] so that the default initialization for the period might be more reasonable.
- try initializing the period to be something close to be what you expect it to be.

**Comment by GStechschulte** [comment]
> A couple of possibilities here, but it's a bit tricky to tell without the data although i can probably mock it if necessary:
> 
> * are you running the GP in single or double precision? try running in double.
> * try also standardizing the inputs to have range [0, 1] so that the default initialization for the period might be more reasonable.
> * try initializing the period to be something close to be what you expect it to be.

Thanks for the quick reply. I have tried a combination of everything you had mentioned, and surprisingly, the issue still persists. I have emailed you regarding more information and with access to my notebook. If we find a solution, I will update this issue thread with the solution to the problem.

**Comment by GStechschulte** [comment]
Coming back to this issue after some time. The reason for the constant prediction value was due to the fact that the composition of kernels was inadequate in extrapolating outside of the training data. 

The data indeed is periodic with some local variations, however, I initially only specified one periodic kernel. When analyzing the autocorrelation plot (ACF), the periodicity occurred at more than one period. For example, if your time series is aggregated at 1 hour intervals, you could use the ACF to identify 12, 24, 48 hour seasonal patterns. You should then normalize your time scale between [0, 1] and identify the range of the seasonal patterns. For example:

```python3
# ACF shows strong autocorrelation for values between 22 and 25 hours
df['t_scaled'].iloc[22], df['t_scaled'].iloc[25]
```
(0.18487394957983194, 0.21008403361344538)

From there, one can specify a period constraint in the periodic kernel to reflect this belief of a daily seasonal pattern

```python3
period_constraint_long = gpytorch.constraints.Interval(0.18, 0.21)

seasonal_periodic_long = gpytorch.kernels.ScaleKernel(
    gpytorch.kernels.PeriodicKernel(
        period_length_constraint=period_constraint_long
        )
    )
```

From here, you can multiply or add other kernels. Hope this helps with anyone modeling seasonal patterns in time series.
